docs/weave/tutorial/11-cron.md

```diff
@@ -0,0 +1,272 @@
+---
+id: cron-tasks
+title: Scheduled Tasks
+sidebar_label: Scheduled Tasks
+---
+
+In almost every real time software, the need to schedule jobs to be executed at a certain time in future is constant. In blockchain context, this could be sending tokens, distributing revenue, scheduling a proposal, basically anything that could be executed as a transaction(a change in state). One of the functionality missing in older blockchain protocols is assuring the execution time of a transaction precisely is impossible due to the [Probabilistic Finality](../basics/02-consensus.md#probabilistic-finality) nature of the consensus algorithms. One of the wonders of Tendermint Consensus engine is we can approximately predict the execution time of a block thanks to [Immediate Finality](../basics/02-consensus.md#immediate-finality).
```

Comment:
What do you think of ``` The early generation of blockchain protocols had a hard time assuring the execution time of a transaction precisely is impossible due to the [Probabilistic Finality](../basics/02-consensus.md#probabilistic-finality) nature of the consensus algorithms```

Reply:
Maybe it is just me, but I still don't understand it.

> In the earlier generations of blockchain protocols it was impossible to use synchronize wall clock value. Instead using a wall clock, ?vector clock? was used.

I don't know if this is describing the same. Maybe @webmaster128 will know better.

Reply:
As far as I can tell, the correctness of timestamps and finality are unrelated. The problem is that (partially because of varying block times), the mining time is not part of the block validation. So a miner can put basically every time in there and the network will accept it as a correct block. This leads to potential attacks where a miner can change the field in a way favourable to them. In Tendermint, the time stamp must be in a very small range in order for other validator's to accept a proposed block as valid.

Those dimensions (finality and timestamp guarantees) are independent, as all of those 4 combinations are possible
1. No instant finality/no timestamp guarantees: e.g. Ethereum
2. Instant finality/no timestamp guarantees: any other instant finality protocol that does not care about block creation timestamps
3. No instant finality/Timestamp guarantees: a protocol with probabilistic finality, that supports forks, but creation time is part of the block validation (e.g. some time range relative to the current time and previous block time)
4. Instant finality/Timestamp guarantees: e.g. Tendermint